### PR TITLE
Milestone 3 - Friendships v2

### PR DESCRIPTION
**Pull request by John-Arboleda**
# Milestone 3 - Friendships v2

This is the third milestone of the Social Media Project from the Ruby on Rails Module. Starting from the previous milestone, the friendship model and users associations were improved, in order to make friendships with 2 rows for mutual friendship.

### CODE REVIEW

File: app/models/friendship.rb [hunk 1]
```diff
@@ -1,4 +1,13 @@
...
+  belongs_to :invitation_sender, class_name: 'User'
+  belongs_to :invitation_receiver, class_name: 'User'
+
+  def add_reverse_friendship_pair
+    friendship = Friendship.find(id)
+    return unless friendship
+
+    Friendship.create(invitation_sender_id: friendship.invitation_sender_id,
+                      invitation_receiver_id: friendship.invitation_receiver_id, status: true)
+  end
```
Comment (bini-i) [inline app/models/friendship.rb]: - [x] It is good to create a reverse record of friendship once it is approved, but kindly make sure to also reverse sender_id and receiver_id in the new record.
```
eg: #1=> sender_id: 1, receiver_id: 2
      #2=> sender_id: 2, receive_id: 1
```

File: app/models/user.rb [hunk 2]
```diff
@@ -9,16 +9,52 @@ class User < ApplicationRecord
...
+  def reject_request(user)
+    friend_request = invitations_received.where(invitation_sender_id: user.id).first
+    friend_request.destroy
+  end
+
+  def friends
+    friend_sender = User.where(id: invitations_sent.where(status: true).pluck(:invitation_receiver_id))
+    friend_receiver = User.where(id: invitations_received.where(status: true).pluck(:invitation_sender_id))
+    friend_sender + friend_receiver
+  end
```
Comment (bini-i) [inline app/models/user.rb]: - [x] Once you update the friendship model to record a reverse of every friendship, you can take advantage of it to get friends of a user by using one column (user_id)  with a status of true. Currently, you are checking both columns(user_id and friend_id), which results in a greater number of queries and performs less efficiently.

**Review by bini-i — CHANGES_REQUESTED** [review]
# Status: Changes Required :yellow_circle:            
### Hello @John-Arboleda  :wave: 

Your project looks good and shows you have put a lot of effort into it. Kindly work on the following suggestions to further improve your project.

## Summary
#### Highlights
- Sending and accepting friend requests works smoothly.
- Posts of user and approved friends can only be seen.
- You have included add_reverse_friendship to establish a two-way friendship.
- Unit and Integration tests are included.

Kindly work on the following suggestions to further improve your project (see inline comments below :point_down: )
## Optional suggestions
-   [_OPTIONAL_]

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

#### If you have any questions feel free to ask within the PR thread.
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (Code Reviewer)
> GitHub: [@bini-i](https://github.com/bini-i) 
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)

**Review by eliasecasta — APPROVED** [review]
## Status:  Approved 🟢

Hi, @John-Arboleda, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**